### PR TITLE
keep track of unresolved value of symbolic-ref in ref iterators

### PR DESCRIPTION
**Pull request by john-cai**
For reftable development, it's useful to be able to print out the direct value of a symbolic reference before resolution. This is currently possible with git-for-each-ref, but since the iterators do not keep track of the value of the symbolic ref, a separate call needs to be made each time making it inefficient.

Address this inefficiency by keeping track of the value of the symbolic reference in the ref iterator. This patch series also ends with a commit to use this value in the iterator through callbacks in ref-filter.c.

This series started with [1] but I decided to send a separate patch series since it is substantially different.

Benchmarking shows that with these changes, we experience a speedup in git-for-each-ref(1) on a repository with many symbolic refs:

$ hyperfine --warmup 5 "git for-each-ref --format='%(refname) %(objectname) %(symref)'" "~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)'"
Benchmark 1: git for-each-ref --format='%(refname) %(objectname) %(symref)'
  Time (mean ± σ):     905.1 ms ±  13.2 ms    [User: 56.3 ms, System: 628.6 ms]
  Range (min … max):   893.4 ms … 936.9 ms    10 runs

Benchmark 2: ~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)'
  Time (mean ± σ):     482.2 ms ±  26.4 ms    [User: 34.7 ms, System: 410.6 ms]
  Range (min … max):   459.4 ms … 541.8 ms    10 runs

Summary
  ~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)' ran
    1.88 ± 0.11 times faster than git for-each-ref --format='%(refname) %(objectname) %(symref)'

Changes since V3:
- Added blank line to separate declarations from statements
- unconditionally set referent member in base iterator in files backend

Changes since V2:
- Style tweaks
- Added a NEEDSWORK comment around fallback code
- Stay consistent in the meaning of a non-NULL referent member in the iterator

Changes since V1
- Use the return value from refs_resolve_ref_unsafe instead of using an out parameter

1. https://lore.kernel.org/git/pull.1684.git.git.1709592718743.gitgitgadget@gmail.com/

cc: Phillip Wood <phillip.wood123@gmail.com>
cc: Kristoffer Haugsbakk <code@khaugsbakk.name>
cc: Jeff King <peff@peff.net>
cc: Patrick Steinhardt <ps@pks.im>
cc: Jean-Noël Avila <avila.jn@gmail.com>
cc: Linus Arver <linusarver@gmail.com>

### CODE REVIEW

**Comment by MXzander** [comment]
> For reftable development, it's useful to be able to print out the direct value of a symbolic reference before resolution. This is currently possible with git-for-each-ref, but since the iterators do not keep track of the value of the symbolic ref, a separate call needs to be made each time making it inefficient.
> 
> Address this inefficiency by keeping track of the value of the symbolic reference in the ref iterator. This patch series also ends with a commit to use this value in the iterator through callbacks in ref-filter.c.
> 
> This series started with [1] but I decided to send a separate patch series since it is substantially different.
> 
> Benchmarking shows that with these changes, we experience a speedup in git-for-each-ref(1) on a repository with many symbolic refs:
> 
> $ hyperfine --warmup 5 "git for-each-ref --format='%(refname) %(objectname) %(symref)'" "~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)'" Benchmark 1: git for-each-ref --format='%(refname) %(objectname) %(symref)' Time (mean ± σ): 905.1 ms ± 13.2 ms [User: 56.3 ms, System: 628.6 ms] Range (min … max): 893.4 ms … 936.9 ms 10 runs
> 
> Benchmark 2: ~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)' Time (mean ± σ): 482.2 ms ± 26.4 ms [User: 34.7 ms, System: 410.6 ms] Range (min … max): 459.4 ms … 541.8 ms 10 runs
> 
> Summary ~/Projects/git/git for-each-ref --format='%(refname) %(objectname) %(symref)' ran 1.88 ± 0.11 times faster than git for-each-ref --format='%(refname) %(objectname) %(symref)'
> 
> Changes since V3:
> 
> * Added blank line to separate declarations from statements
> * unconditionally set referent member in base iterator in files backend
> 
> Changes since V2:
> 
> * Style tweaks
> * Added a NEEDSWORK comment around fallback code
> * Stay consistent in the meaning of a non-NULL referent member in the iterator
> 
> Changes since V1
> 
> * Use the return value from refs_resolve_ref_unsafe instead of using an out parameter
> 
> 1. https://lore.kernel.org/git/pull.1684.git.git.1709592718743.gitgitgadget@gmail.com/
> 
> cc: Phillip Wood [phillip.wood123@gmail.com](mailto:phillip.wood123@gmail.com) cc: Kristoffer Haugsbakk [code@khaugsbakk.name](mailto:code@khaugsbakk.name) cc: Jeff King [peff@peff.net](mailto:peff@peff.net) cc: Patrick Steinhardt [ps@pks.im](mailto:ps@pks.im) cc: Jean-Noël Avila [avila.jn@gmail.com](mailto:avila.jn@gmail.com) cc: Linus Arver [linusarver@gmail.com](mailto:linusarver@gmail.com) cc: shejialuo [shejialuo@gmail.com](mailto:shejialuo@gmail.com)

File: refs/files-backend.c [hunk 1]
```diff
@@ -245,9 +245,12 @@ static void loose_fill_ref_dir_regular_file(struct files_ref_store *refs,
 {
```
[the remark this answers is not in the document]
Comment (mfdonline) [reply refs/files-backend.c:245]: Root

**Pull request closed without merging**